uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+        self.stdout.write(
+            f"Executing: {self.primary_key_constraint['temp_constraint_create_statement']}"
+        )
+        self.cursor.execute(
+            self.primary_key_constraint["temp_constraint_create_statement"]
+        )
+
+    def get_foreign_key_constraints(self):
+        fkey_sql = f"""
+            SELECT conname AS constraint_name, confrelid::regclass::text AS refrenced_table, pg_get_constraintdef(oid), conrelid::regclass::text AS referencing_table
+            FROM pg_constraint 
+            WHERE contype = 'f'
+                AND (
+                    conrelid = '{self.table_name}'::regclass 
+                    OR
+                    confrelid = '{self.table_name}'::regclass
+                )      
+        """
+        self.cursor.execute(fkey_sql)
+        results = self.cursor.fetchall()
+
+        self.stdout.write(
+            f"Found {len(results)} foreign key constraints, where {self.table_name} is the referencing or referenced table"
+        )
+
+        fk_constraints = []
+
+        for row in results:
+            constraint_name = row[0]
+            # referenced_table_name = row[1]
+            constraintdef = row[2]
+            referencing_table = row[3]
+
+            fk_constraints.append(
+                {
+                    "constraint_name": constraint_name,
+                    "create_statement": f"ALTER TABLE {referencing_table} ADD CONSTRAINT {constraint_name} {constraintdef} NOT VALID",
+                    "delete_statement": f"ALTER TABLE {referencing_table} DROP CONSTRAINT IF EXISTS {constraint_name}",
+                    "validate_statement": f"ALTER TABLE {referencing_table} VALIDATE CONSTRAINT {constraint_name}",
+                }
+            )
+
+        return fk_constraints
+
+    def drop_foreign_keys(self):
+        self.stdout.write("Dropping foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['delete_statement']}")
+            self.cursor.execute(constraint["delete_statement"])
+
+    def add_foreign_keys(self):
+        self.stdout.write("Creating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            try:
+                self.stdout.write(
+                    f"Executing: {constraint['create_statement']}"
+                )
+                self.cursor.execute(constraint["create_statement"])
+            except (
+                ProgrammingError
+            ):  # better if psycopg2.errors.DuplicateObject:
+                self.stdout.write(
+                    f"{constraint['constraint_name']} already exists."
+                )
+
+    def validate_foreign_keys(self):
+        self.stdout.write("Validating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['validate_statement']}")
+            self.cursor.execute(constraint["validate_statement"])
```

Comment:
What happens if one of these statements fails? Does it throw?

Reply:
TODO: Double check this.

Reply:
In general.. where we are issuing a `self.cursor.execute("RAW SQL")`, do we actually need to be checking the return values and throwing to ensure we roll back if bad stuff happens?

Reply:
Have checked this throws. Haven't yet written a test for it.